File: Data Structures & Algorithms/balanced-binary-tree/submission-0.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {boolean}
     */
    isBalanced(root) {
        if(!root) return true; 

        const getHeight = (root) => {
            if(!root) return 0;

            return Math.max(getHeight(root.left), getHeight(root.right)) + 1;
        }

        let left = getHeight(root.left);
        let right = getHeight(root.right);

        if(Math.abs(left - right) > 1) return false;

        return (Math.abs(left - right) <= 1 && this.isBalanced(root.left) && this.isBalanced(root.right));
    }
}

File: Data Structures & Algorithms/insert-into-a-binary-search-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def insertIntoBST(self, root: Optional[TreeNode], val: int) -> Optional[TreeNode]:
        def insertNode(curr, val):
            if not curr:
                return TreeNode(val)

            if val < curr.val:
                curr.left = insertNode(curr.left, val)
            elif val > curr.val:
                curr.right = insertNode(curr.right, val)
            return curr
            
        return insertNode(root, val)